scaletest/trafficgen/run.go

```diff
@@ -0,0 +1,274 @@
+	rch := make(chan error)
+	wch := make(chan error)
+
+	go func() {
+		<-deadlineCtx.Done()
+		logger.Debug(ctx, "context deadline reached", slog.F("duration", time.Since(start)))
+	}()
+
+	// Read forever in the background.
+	go func() {
+		logger.Debug(ctx, "reading from agent", slog.F("agent_id", agentID))
+		rch <- drainContext(deadlineCtx, &crw, bytesPerTick*2)
+		logger.Debug(ctx, "done reading from agent", slog.F("agent_id", agentID))
+		conn.Close()
+		close(rch)
+	}()
+
+	// Write random data to the PTY every tick.
+	go func() {
+		logger.Debug(ctx, "writing to agent", slog.F("agent_id", agentID))
+		wch <- writeRandomData(deadlineCtx, &crw, bytesPerTick, tick.C)
+		logger.Debug(ctx, "done writing to agent", slog.F("agent_id", agentID))
+		close(wch)
+	}()
+
+	// Write until the context is canceled.
+	if wErr := <-wch; wErr != nil {
+		return xerrors.Errorf("write to pty: %w", wErr)
+	}
+	if rErr := <-rch; rErr != nil {
+		return xerrors.Errorf("read from pty: %w", rErr)
+	}
+
+	duration := time.Since(start)
+
+	logger.Info(ctx, "results",
+		slog.F("duration", duration),
+		slog.F("sent", crw.BytesWritten()),
+		slog.F("rcvd", crw.BytesRead()),
+	)
+
+	return nil
+}
+
+// Cleanup does nothing, successfully.
+func (*Runner) Cleanup(context.Context, string) error {
+	return nil
+}
+
+// drainContext drains from src until it returns io.EOF or ctx times out.
+func drainContext(ctx context.Context, src io.Reader, bufSize int64) error {
+	errCh := make(chan error, 1)
+	done := make(chan struct{})
+	go func() {
+		tmp := make([]byte, bufSize)
+		buf := bytes.NewBuffer(tmp)
+		for {
+			select {
+			case <-done:
+				return
+			default:
+				_, err := io.CopyN(buf, src, 1)
+				if ctx.Err() != nil {
+					return // context canceled while we were copying.
+				}
+				if err != nil {
+					errCh <- err
+					close(errCh)
+					return
+				}
+			}
+		}
+	}()
+	for {
+		select {
+		case <-ctx.Done():
+			close(done)
+			return nil
+		case err := <-errCh:
+			if err != nil {
+				if xerrors.Is(err, io.EOF) {
+					return nil
+				}
+				// It's OK if the context is canceled.
+				if xerrors.Is(err, context.DeadlineExceeded) {
+					return nil
+				}
+				return err
+			}
+		}
+	}
+}
+
+func writeRandomData(ctx context.Context, dst io.Writer, size int64, tick <-chan time.Time) error {
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-tick:
+			payload := "#" + mustRandStr(size-1)
+			data, err := json.Marshal(codersdk.ReconnectingPTYRequest{
+				Data: payload,
+			})
+			if err != nil {
+				return err
+			}
+			if _, err := copyContext(ctx, dst, data); err != nil {
+				return err
+			}
+		}
+	}
+}
+
+// copyContext copies from src to dst until ctx is canceled.
+func copyContext(ctx context.Context, dst io.Writer, src []byte) (int, error) {
+	var count int
+	for {
+		select {
+		case <-ctx.Done():
+			return count, nil
+		default:
+			for idx := range src {
+				n, err := dst.Write(src[idx : idx+1])
```

Comment:
Why do we write a single char at a time? Without also setting a write deadline, we can't really ensure we exit on timeout (might hang writing due to network hiccup etc), and context cancellation won't propagate in either case. For that to happen the writer needs to be closed.

Reply:
While I was testing with a slow connection, I found that writing one byte at a time allowed detecting the context cancellation more quickly. Closing the writer at the deadline does definitely help here, silly oversight on my part!